src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+	bool overcommit;
+} dispatch_workqueue_s;
+typedef dispatch_workqueue_s *dispatch_workqueue_t;
+
+/*
+ * Overcommit queues.
+ *
+ * The overcommit pool uses a simple coarse locking policy:
+ *   all overcommit pool operations are performed while
+ *   holding the mutex.
+ *
+ * This enables a simple, but non-scalable, implementation
+ * that is adequate as long as operations on the overcommit
+ * pool are relatively infrequent.
+ */
+typedef struct {
+	uint mask;
+	int num_spares;
+	dispatch_workqueue_t queues[WORKQ_NUM_PRIORITIES];
+	pthread_mutex_t mutex;
+	pthread_cond_t spare_workers;
+} dispatch_overcommit_pool_s;
+typedef dispatch_overcommit_pool_s * dispatch_overcommit_pool_t;
+
+static dispatch_overcommit_pool_s overcommit_pool;
+
+/*
+ * Normal queues.
+ *
+ * The normal pool uses a combination of atomic
+ * operations on a volatile mask word that encodes which
+ * workqueues are non-empty and per-queue synchronization
+ * (currently implemented via pthread_spin_locks).
+ * This is slightly more scalable than the overcommit queue, but
+ * is may be an area that needs future performance enhancement.
+ */
+typedef struct {
+	volatile uint mask;
+	dispatch_workqueue_t queues[WORKQ_NUM_PRIORITIES];
+} dispatch_pool_s;
+typedef dispatch_pool_s * dispatch_pool_t;
+
+static dispatch_pool_s normal_pool;
+
+
+/*
+ * internal prototypes
+ */
+static void	_dispatch_spawn_thread(void *(start_func)(void*));
+static void* _dispatch_overcommit_worker_main(void* unused);
+static void* _dispatch_normal_worker_main(void* unused);
+static void* _dispatch_thread_monitor_main(void* unused);
+
+// TODO: declare in queue_internal.h
+extern int _dispatch_pthread_sigmask(int how, sigset_t *set, sigset_t *oset);
+
+
+#define WORKITEM_WORKER_SHOULD_WAIT ((void*)1)
+#define WORKITEM_WORKER_SHOULD_EXIT ((void*)2)
+
+static dispatch_workitem_t
+_dispatch_alloc_workitem(void (*workitem_func)(void *), void * workitem_arg)
+{
+	dispatch_workitem_t wi = _dispatch_calloc(1, sizeof(struct work));
+	wi->func = workitem_func;
+	wi->func_arg = workitem_arg;
+	return wi;
+}
+
+static void
+_dispatch_dealloc_workitem(dispatch_workitem_t wi)
+{
+	free(wi);
+}
+
+
+static void
+_dispatch_allocate_workqueues(dispatch_workqueue_t *queues, int num_queues, bool overcommit)
+{
+	int r;
+
+	for (int i=0; i<num_queues; i++) {
+		dispatch_workqueue_t wq;
+		wq = _dispatch_calloc(1, ROUND_UP_TO_CACHELINE_SIZE(sizeof(*wq)));
+
+		wq->priority = i;
+		wq->overcommit = overcommit;
+		STAILQ_INIT(&wq->item_listhead);
+		r = pthread_spin_init(&wq->sl, PTHREAD_PROCESS_PRIVATE);
+		(void)dispatch_assume_zero(r);
+
+		queues[i] = wq;
+	}
+}
+
+static dispatch_once_t _dispatch_workqueue_init_once_pred;
+
+static void
+_dispatch_workqueue_init_once(void *context DISPATCH_UNUSED)
+{
+	int r;
+
+	// Initialize thread pool size parameters
+	int32_t num_cpus = dispatch_hw_config(active_cpus);
+	int32_t low_target = num_cpus;
+	char* env_low_target = getenv("LIBDISPATCH_THREADPOOL_LOW_TARGET");
```

Comment:
same that below, use const, and the star sticks to the variable name not the type